plaidml/bridge/openvino/ops/detection_output.cpp

```diff
@@ -0,0 +1,223 @@
+// Copyright (C) 2021 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include "plaidml_ops.hpp"
+
+#include "ngraph/opsets/opset.hpp"
+#include "ngraph/opsets/opset4.hpp"
+
+#include "plaidml/op/op.h"
+
+using namespace plaidml;          // NOLINT[build/namespaces]
+using namespace InferenceEngine;  // NOLINT[build/namespaces]
+using namespace plaidml::edsl;
+
+namespace PlaidMLPlugin {
+
+edsl::Tensor DecodeBoxes(edsl::Tensor priors, edsl::Tensor prior_variances, edsl::Tensor location,
```

Comment:
> Part of this function appears fairly similar to the center_point_box parameter in the OpLib NMS.

Yes there are two types of box encoding in both functions, namely `corner` and `center`. In both functions they are doing pretty much the same thing but there are some differences, too. In terms of feasibility, I think merging these two functions is not complicated. There are a few differences we need to consider though.

- The different outputs of `NonMaxSuppression` and `DetectionOputput`. That is, the output of `NonMaxSuppression` is in the form of

```
[batch_index, class_index, box_score]
```
while the other one is
```
[batch_index, class_index, box_score, x_min, y_min, x_max, y_max]
```
If we merge `decodeBoxes` into OpLib then we should have two different types of output based on something like decode_mode. I'm not sure if this is acceptable/preferred.

- When there are 5 inputs of `DetectionOutput`, at [line 151](https://github.com/plaidml/plaidml/pull/1741/files#diff-1a9d0f648e4a5a282f62d552d60faaa7fb700552308e9d569c71d27eea535d75R151) we have to decode twice to get `decoded_boxes` (first with `ArmLocation` and then `Location`). Of course we can pass `ArmLocation` into OpLib NMS and do decoding twice inside of NMS but this seems to be a detection-output-specific parameter. It seems to me `decodeBoxes` should be a relatively standalone fuction.

What do you think?

Reply:
Thanks for that explanation. In my opinion, it would be make sense to have 3 separate parameters determining:
* Corner vs. Center
* Output format
* Whether additional ARM prediction inputs are used

We wouldn't be using all 8 possible combinations of these 3 parameters -- at least not yet -- so it would be reasonable to just throw if an unsupported/untested choice of parameters is used. But in my opinion these 3 parameters are not intrinsically correlated, and so I think a design separating them is preferable unless there's some reason that separating them would be hard to implement.

Additionally, I think having a `decodeBoxes` helper function is the oplib is reasonable. I think at the moment this would only be used by NMS, and at the moment I can't think of other ops that would use it. But, it's not inherently linked to NMS, so it's reasonable to make it available in case someone else does have a use for it.

Reply:
Basically done. Please check.